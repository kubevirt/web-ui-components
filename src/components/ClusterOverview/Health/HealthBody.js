import React from 'react';
import PropTypes from 'prop-types';
import { Icon } from 'patternfly-react';

const healtyIcon = (
  <div className="kubevirt-health__icon--ok kubevirt-health__icon">
    <Icon type="fa" name="check-circle" />
  </div>
);

const errorIcon = (
  <div className="kubevirt-health__icon--error kubevirt-health__icon">
    <Icon type="fa" name="exclamation-circle" />
  </div>
);

const HealthBody = ({ data }) => (
  <React.Fragment>
    {data.healthy ? healtyIcon : errorIcon}
    {data.message}
  </React.Fragment>
);

HealthBody.propTypes = {
  data: PropTypes.object.isRequired,
};

export default HealthBody;
