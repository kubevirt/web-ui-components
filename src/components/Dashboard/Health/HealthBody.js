import React from 'react';
import PropTypes from 'prop-types';

export const HealthBody = ({ children }) => <div className="kubevirt-health__body">{children}</div>;

HealthBody.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
};
