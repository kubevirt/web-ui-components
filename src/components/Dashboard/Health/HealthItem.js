import React from 'react';
import PropTypes from 'prop-types';
import { Icon } from 'patternfly-react';

import { InlineLoading } from '../../Loading';

export const OK_STATE = 'OK_STATE';
export const ERROR_STATE = 'ERROR_STATE';
export const WARNING_STATE = 'WARNING_STATE';
export const LOADING_STATE = 'LOADING_STATE';

const getIcon = state => {
  let icon;
  let className;
  switch (state) {
    case OK_STATE:
      icon = 'check-circle';
      className = 'ok';
      break;
    case ERROR_STATE:
      icon = 'exclamation-circle';
      className = 'error';
      break;
    case WARNING_STATE:
    default:
      icon = 'exclamation-triangle';
      className = 'warning';
  }
  return (
    <div className={`kubevirt-health__icon kubevirt-health__icon--${className}`}>
      <Icon type="fa" name={icon} />
    </div>
  );
};

export const HealthItem = ({ state, LoadingComponent, message, details }) => (
  <div className="kubevirt-health__item">
    {state === LOADING_STATE ? <LoadingComponent /> : getIcon(state)}
    <div>
      <span className="kubevirt-health__text">{message}</span>
      {details && <div className="kubevirt-health__text kubevirt-health__subtitle">{details}</div>}
    </div>
  </div>
);

HealthItem.defaultProps = {
  details: null,
  state: null,
  LoadingComponent: InlineLoading,
};

HealthItem.propTypes = {
  message: PropTypes.string.isRequired,
  details: PropTypes.string,
  state: PropTypes.string,
  LoadingComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
};
