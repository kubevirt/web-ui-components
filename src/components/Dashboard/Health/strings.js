export const OCS_HEALTHY = 'OCS is healthy';
export const OCS_DEGRADED = 'OCS health is degraded';
export const OCS_ERROR = 'OCS health is in error state';
export const OCS_UNKNOWN = 'OCS is not available';

export const CNV_HEALTHY = 'CNV is healthy';
export const CNV_ERROR = 'CNV is in error state';

export const OCP_HEALTHY = 'OpenShift is healthy';
export const OCP_ERROR = 'Openshift is in error state';
