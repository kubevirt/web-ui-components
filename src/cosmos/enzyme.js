import createContext from 'react-cosmos-test/enzyme';

import paths from '../../config/paths';

export default function createTestContext(args) {
  args.proxies = args.proxies || []; // workaround of react-cosmos-test bug
  return createContext({
    ...args,
    cosmosConfigPath: `${paths.config}/cosmos.config.js`,
  });
}
