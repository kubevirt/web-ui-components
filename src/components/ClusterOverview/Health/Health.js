import React from 'react';
import PropTypes from 'prop-types';

import {
  DashboardCard,
  DashboardCardBody,
  DashboardCardHeader,
  DashboardCardTitle,
} from '../../Dashboard/DashboardCard';
import HealthBody from './HealthBody';
import { ClusterOverviewContextGenericConsumer } from '../ClusterOverviewContext';

export const Health = ({ data, loaded }) => (
  <DashboardCard>
    <DashboardCardHeader>
      <DashboardCardTitle>Cluster Health</DashboardCardTitle>
    </DashboardCardHeader>
    <DashboardCardBody className="kubevirt-health__body" isLoading={!loaded}>
      <HealthBody data={data} />
    </DashboardCardBody>
  </DashboardCard>
);

Health.defaultProps = {
  loaded: false,
};

Health.propTypes = {
  data: PropTypes.object.isRequired,
  loaded: PropTypes.bool,
};

const HealthConnected = () => <ClusterOverviewContextGenericConsumer Component={Health} dataPath="healthData" />;

export default HealthConnected;
