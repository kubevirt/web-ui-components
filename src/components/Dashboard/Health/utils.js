import { get } from 'lodash';

import { LOADING_STATE, OK_STATE, ERROR_STATE, WARNING_STATE } from './HealthItem';
import {
  OCS_HEALTHY,
  OCS_DEGRADED,
  OCS_ERROR,
  OCS_UNKNOWN,
  OCP_HEALTHY,
  OCP_ERROR,
  CNV_HEALTHY,
  CNV_ERROR,
} from './strings';

export const getKubevirtHealthState = kubevirtHealth => {
  if (!kubevirtHealth) {
    return { state: LOADING_STATE };
  }
  return get(kubevirtHealth, 'apiserver.connectivity') === 'ok'
    ? { message: CNV_HEALTHY, state: OK_STATE }
    : { message: CNV_ERROR, state: ERROR_STATE };
};

export const getK8sHealthState = k8sHealth => {
  if (!k8sHealth) {
    return { state: LOADING_STATE };
  }
  return get(k8sHealth, 'response') === 'ok'
    ? { message: OCP_HEALTHY, state: OK_STATE }
    : { message: OCP_ERROR, state: ERROR_STATE };
};

const OCSHealthStatus = {
  0: {
    message: OCS_HEALTHY,
    state: OK_STATE,
  },
  1: {
    message: OCS_DEGRADED,
    state: WARNING_STATE,
  },
  2: {
    message: OCS_ERROR,
    state: ERROR_STATE,
  },
  3: {
    message: OCS_UNKNOWN,
    state: ERROR_STATE,
  },
};

export const getOCSHealthState = ocsResponse => {
  if (!ocsResponse) {
    return { state: LOADING_STATE };
  }
  const value = get(ocsResponse, 'result[0].value[1]');
  return OCSHealthStatus[value] || OCSHealthStatus[3];
};
