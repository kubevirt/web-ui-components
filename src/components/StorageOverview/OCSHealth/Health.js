import React from 'react';
import PropTypes from 'prop-types';

import {
  DashboardCard,
  DashboardCardBody,
  DashboardCardHeader,
  DashboardCardTitle,
} from '../../Dashboard/DashboardCard';
import { HealthItem, LOADING_STATE } from '../../Dashboard/Health/HealthItem';
import { HealthBody } from '../../Dashboard/Health/HealthBody';
import { InlineLoading } from '../../Loading';
import { getOCSHealthState } from '../../Dashboard/Health/utils';
import { StorageOverviewContext } from '../StorageOverviewContext';

export const OCSHealth = ({ ocsHealthResponse, LoadingComponent }) => {
  const state = getOCSHealthState(ocsHealthResponse);
  return (
    <DashboardCard>
      <DashboardCardHeader>
        <DashboardCardTitle>Health</DashboardCardTitle>
      </DashboardCardHeader>
      <DashboardCardBody isLoading={state.state === LOADING_STATE} LoadingComponent={LoadingComponent}>
        <HealthBody>
          <HealthItem message={state.message} state={state.state} />
        </HealthBody>
      </DashboardCardBody>
    </DashboardCard>
  );
};

OCSHealth.defaultProps = {
  ocsHealthResponse: null,
  LoadingComponent: InlineLoading,
};

OCSHealth.propTypes = {
  ocsHealthResponse: PropTypes.object,
  LoadingComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
};

export const OCSHealthConnected = () => (
  <StorageOverviewContext.Consumer>{props => <OCSHealth {...props} />}</StorageOverviewContext.Consumer>
);
