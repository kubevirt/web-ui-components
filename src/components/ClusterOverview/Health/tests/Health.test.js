import React from 'react';
import { render } from 'enzyme';

import { Health } from '../Health';
import { default as HealthFixtures } from '../fixtures/Health.fixture';

const testHealthOverview = () => <Health {...HealthFixtures[0].props} />;

describe('<Health />', () => {
  it('renders correctly', () => {
    const component = render(testHealthOverview());
    expect(component).toMatchSnapshot();
  });
});
