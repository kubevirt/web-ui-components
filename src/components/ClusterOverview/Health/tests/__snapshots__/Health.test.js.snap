// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Health /> renders correctly 1`] = `
<div
  class="card-pf kubevirt-dashboard__card"
>
  <div
    class="card-pf-heading kubevirt-dashboard__card-heading"
  >
    <h2
      class="card-pf-title"
    >
      Cluster Health
    </h2>
  </div>
  <div
    class="card-pf-body kubevirt-health__body"
  >
    <div
      class="kubevirt-health__icon--error"
    >
      <span
        aria-hidden="true"
        class="fa fa-exclamation-circle"
      />
    </div>
    Error message
  </div>
</div>
`;
